# `pandas`

This workshop's goal&mdash;which is facilitated by this Jupyter notebook&mdash;is to give attendees the confidence to use `pandas` in their research projects. Basic familiarity with Python *is* assumed.

`pandas` is designed to make it easier to work with structured data. Most of the analyses you might perform will likely involve using tabular data, e.g., from .csv files or relational databases (e.g., SQL). The `DataFrame` object in `pandas` is "a two-dimensional tabular, column-oriented data structure with both row and column labels."

If you're curious:

>The `pandas` name itself is derived from *panel data*, an econometrics term for multidimensional structured data sets, and *Python data analysis* itself. After getting introduced, you can consult the full [`pandas` documentation](http://pandas.pydata.org/pandas-docs/stable/).

To motivate this workshop, we'll work with example data and go through the various steps you might need to prepare data for analysis. You'll (hopefully) realize that doing this type of work is much more difficult using Python's built-in data structures.

### Table of Contents

1 - [The DataFrame](#section1)<br>

2 - [Rename, Index, and Slice](#section2)<br>

3 - [Manipulating Columns](#section3)<br>

4 - [Merging](#section4)<br>

5 - [Calculating Unique And Missing Values](#section5)<br>

6 - [Groupby](#section6)<br>

7- [Exporting To CSV](#section7)<br>

8 - [Handling Missing Values (Boolean Indexing)](#section8)<br>

9 - [Sorting Values](#section9)<br>

10 - [Plotting In Pandas](#section10)<br>

## 1. The DataFrame: Importing Data and Summary Statistics <a id="section1"/>
The data used in these examples is available in the following [GitHub repository](https://github.com/dlab-berkeley/introduction-to-pandas). If you've [cloned that repo](https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-clone), which is the recommended approach, you'll have everything you need to run this notebook. Otherwise, you can download the data file(s) from the above link. (Note: this notebook assumes that the data files are in a directory named `data/` found within your current working directory.)

For this example, we're working with European unemployment data from Eurostat, which is hosted by [Google](https://code.google.com/p/dspl/downloads/list). There are several .csv files that we'll work with in this workshop.

Let's begin by importing `pandas` using the conventional abbreviation.

In [1]:
%matplotlib inline

import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt

mpl.rc('savefig', dpi=200)
plt.style.use('ggplot')
plt.rcParams['xtick.minor.size'] = 0
plt.rcParams['ytick.minor.size'] = 0

In [2]:
pd.__version__

'0.23.4'

The `read_csv()` function in `pandas` allows us to easily import our data. By default, it assumes the data is comma-delimited. However, you can specify the delimiter used in your data (e.g., tab, semicolon, pipe, etc.). There are several parameters that you can specify. See the documentation [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). `read_csv()` returns a `DataFrame`.

Notice that we call `read_csv()` using the `pd` abbreviation from the import statement above.

In [3]:
!tree

.
├── 2015-dlab-introduction-workshop.ipynb
├── 2015-dlab-introduction-workshop-solutions.ipynb
├── 2016-kuleshov-python-numpy-tutorial.ipynb
├── 2017-fonnesbeck-1_Data_Preparation.ipynb
├── 2017-fonnesbeck-2_Basic_Bayesian_Inference.ipynb
├── 2017-fonnesbeck-3-Fitting_Regression_Models.ipynb
├── 2017-fonnesbeck-4_Dealing_with_Missing_Data.ipynb
├── 2019-lunacek-1_getting_started.ipynb
├── 2019-lunacek-2_groupby.ipynb
├── 2019-lunacek-3_overlay.ipynb
├── 2019-lunacek-4_concat.ipynb
├── 2019-lunacek-5_apply.ipynb
├── 2019-lunacek-6_reshaping.ipynb
├── 2019-lunacek-7_multi_index.ipynb
├── data
│   ├── 2019-12-17-124840-catalog.json
│   ├── _age_groups.csv
│   ├── comfort_data.csv.gz
│   ├── countries.csv
│   ├── country_age.csv
│   ├── country_group_age.csv
│   ├── _country_groups.csv
│   ├── country_group_sex_age.csv
│   ├── country_group_sex.csv
│   ├── country_group_total.csv
│   ├── country_sex_age.csv
│   ├── country_sex.csv
│   ├── country_total.csv
│   ├── ebola
│   │   ├── guinea

In [4]:
unemployment = pd.read_csv('data/country_total.csv')

In [5]:
#pd.read_csv?

In [6]:
images = pd.read_json('data/2019-12-17-124840-catalog.json')

In [7]:
#pd.read_json?

In [8]:
images.head()

archive.host_country                                      archive.notes  \
0                  USA  Images available via API at https://catalog.ar...   
1                  USA  Images available via API at https://catalog.ar...   
2                  USA  Images available via API at https://catalog.ar...   
3                  USA  Images available via API at https://catalog.ar...   
4                  USA  Images available via API at https://catalog.ar...   

  document.accession_to_archive_date document.contact_person  \
0                         2016-08-19              Kevin Wood   
1                         2016-08-19              Kevin Wood   
2                         2016-08-19              Kevin Wood   
3                         2016-08-19              Kevin Wood   
4                         2016-08-19              Kevin Wood   

   document.id_within_archive document.id_within_archive_type document.notes  \
0                    38547962                         NARA ID                  
1                    38547962                         NARA ID                  
2                    38547962                         NARA ID                  
3                    38547962                         NARA ID                  
4                    38547962                         NARA ID                  

  document.record_type document.rights_statement  \
0          ships' logs         CC0 Public Domain   
1          ships' logs         CC0 Public Domain   
2          ships' logs         CC0 Public Domain   
3          ships' logs         CC0 Public Domain   
4          ships' logs         CC0 Public Domain   

  document.standardized_region_list document.start_date  \
0                    north_atlantic          1957-09-30   
1                    north_atlantic          1957-09-30   
2                    north_atlantic          1957-09-30   
3                    north_atlantic          1957-09-30   
4                    north_atlantic          1957-09-30   

                                           file_path  media_type  \
0                       2019-12-16/data/metadata.csv  text/plain   
1  2019-12-16/data/nara_id_storis-wmec-38-1957-lo...  text/plain   
2  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   
3  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   
4  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   

  platform.name                              uuid  
0   USCG Storis                               NaN  
1   USCG Storis                               NaN  
2   USCG Storis  37ea9580210611eaba89ac1f6bab1d82  
3   USCG Storis  3802945c210611eaba89ac1f6bab1d82  
4   USCG Storis  38198212210611eaba89ac1f6bab1d82

Great! You've created a `pandas` `DataFrame`. We can look at our data by using the `.head()` method. By default, this shows the header (column names) and the first five rows. Passing an integer, $n$, to `.head()` returns that number of rows. To see the last $n$ rows, use `.tail()`.

In [9]:
unemployment.head()

country seasonality    month  unemployment  unemployment_rate
0      at         nsa  1993.01        171000                4.5
1      at         nsa  1993.02        175000                4.6
2      at         nsa  1993.03        166000                4.4
3      at         nsa  1993.04        157000                4.1
4      at         nsa  1993.05        147000                3.9

In [10]:
unemployment.tail()

country seasonality    month  unemployment  unemployment_rate
20791      uk       trend  2010.06       2429000                7.7
20792      uk       trend  2010.07       2422000                7.7
20793      uk       trend  2010.08       2429000                7.7
20794      uk       trend  2010.09       2447000                7.8
20795      uk       trend  2010.10       2455000                7.8

To find the number of rows, you can use the `len()` function. Alternatively, you can use the `shape` attribute.

In [11]:
unemployment.shape

(20796, 5)

There are 20,796 rows and 5 columns.

A useful method that generates various summary statistics is `.describe()`.

In [12]:
unemployment.describe()

month  unemployment  unemployment_rate
count  20796.000000  2.079600e+04       19851.000000
mean    1999.401290  7.900818e+05           8.179764
std        7.483751  1.015280e+06           3.922533
min     1983.010000  2.000000e+03           1.100000
25%     1994.090000  1.400000e+05           5.200000
50%     2001.010000  3.100000e+05           7.600000
75%     2006.010000  1.262250e+06          10.000000
max     2010.120000  4.773000e+06          20.900000

You may have noticed that the "count" is lower for the unemployment rate. This is because the summary statistics are based on *non-missing* values and count reflects that.

The values depend on what it's called on. If the `DataFrame` includes both numeric and object (e.g., strings) `dtype`s, it will default to summarizing the numeric data. If `.describe()` is called on strings, for example, it will return the count, number of unique values, and the most frequent value along with its count.

In [13]:
import numpy as np
d = pd.DataFrame(np.array(['Colton', 'Caleb', 'Samuel', 'Caleb']))
d.describe()

0
count       4
unique      3
top     Caleb
freq        2

In [14]:
images.shape

(22, 15)

In [15]:
images.describe()

document.id_within_archive
count                        22.0
mean                   38547962.0
std                           0.0
min                    38547962.0
25%                    38547962.0
50%                    38547962.0
75%                    38547962.0
max                    38547962.0

### Challenge 1: Import Data From A URL

Above, we imported the unemployment data using the `read_csv` function and a relative file path. `read_csv` is [a very flexible method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.read_csv.html); it also allows us to import data using a URL as the file path. 

A csv file with data on world countries and their abbreviations is located at [https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv](https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv) (saved as a string variable `countries_url` below).

Using `read_csv`, import the country data and save it to the variable `countries`.

In [16]:
countries_url = 'https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv'
countries = pd.read_csv(countries_url)
countries.head()

country google_country_code country_group   name_en   name_fr     name_de  \
0      at                  AT            eu   Austria  Autriche  Österreich   
1      be                  BE            eu   Belgium  Belgique     Belgien   
2      bg                  BG            eu  Bulgaria  Bulgarie   Bulgarien   
3      hr                  HR        non-eu   Croatia   Croatie    Kroatien   
4      cy                  CY            eu    Cyprus    Chypre      Zypern   

    latitude  longitude  
0  47.696554  13.345980  
1  50.501045   4.476674  
2  42.725674  25.482322  
3  44.746643  15.340844  
4  35.129141  33.428682

In [17]:
catalog_url = 'https://raw.githubusercontent.com/coltongrainger/rda-image-archive/master/tests/2019-12-16/out/2019-12-17-124840-catalog.json'
catalog = pd.read_json(catalog_url)
images.equals(catalog)

True

### Challenge 2: The `tail` method

DataFrames all have a method called `tail` that takes an integer as an argument and returns a new DataFrame. Before using `tail`, can you guess at what it does? Try using `tail`; was your guess correct?

In [18]:
# use the tail method on either the unemployment or countries dataframe
countries.head()

country google_country_code country_group   name_en   name_fr     name_de  \
0      at                  AT            eu   Austria  Autriche  Österreich   
1      be                  BE            eu   Belgium  Belgique     Belgien   
2      bg                  BG            eu  Bulgaria  Bulgarie   Bulgarien   
3      hr                  HR        non-eu   Croatia   Croatie    Kroatien   
4      cy                  CY            eu    Cyprus    Chypre      Zypern   

    latitude  longitude  
0  47.696554  13.345980  
1  50.501045   4.476674  
2  42.725674  25.482322  
3  44.746643  15.340844  
4  35.129141  33.428682

In [19]:
countries.tail()

country google_country_code country_group         name_en      name_fr  \
25      si                  SI            eu        Slovenia     Slovénie   
26      es                  ES            eu           Spain      Espagne   
27      se                  SE            eu          Sweden        Suède   
28      tr                  TR        non-eu          Turkey      Turquie   
29      uk                  GB            eu  United Kingdom  Royaume-Uni   

                   name_de   latitude  longitude  
25               Slowenien  46.149259  14.986617  
26                 Spanien  39.895013  -2.988296  
27                Schweden  62.198468  14.896307  
28                  Türkei  38.952942  35.439795  
29  Vereinigtes Königreich  54.315447  -2.232612

### Challenge 3: Describe `countries`

It's important to understand a few fundamentals about your data before you start work with it, including what information it contains, how large it is, and how the values are generally distributed.

Using the methods and attributes above, answer the following questions about `countries`:
* what columns does it contain?
* what does each row stand for?
* how many rows and columns does it contain?
* are there any missing values in the latitude or longitude columns? 

Hint: the `head` and `describe` functions, as well as the `shape` attribute, will be helpful here.

In [20]:
images.sample()

archive.host_country                                      archive.notes  \
9                  USA  Images available via API at https://catalog.ar...   

  document.accession_to_archive_date document.contact_person  \
9                         2016-08-19              Kevin Wood   

   document.id_within_archive document.id_within_archive_type document.notes  \
9                    38547962                         NARA ID                  

  document.record_type document.rights_statement  \
9          ships' logs         CC0 Public Domain   

  document.standardized_region_list document.start_date  \
9                    north_atlantic          1957-09-30   

                                           file_path  media_type  \
9  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   

  platform.name                              uuid  
9   USCG Storis  38923ccc210611eaba89ac1f6bab1d82

In [21]:
images.columns

Index(['archive.host_country', 'archive.notes',
       'document.accession_to_archive_date', 'document.contact_person',
       'document.id_within_archive', 'document.id_within_archive_type',
       'document.notes', 'document.record_type', 'document.rights_statement',
       'document.standardized_region_list', 'document.start_date', 'file_path',
       'media_type', 'platform.name', 'uuid'],
      dtype='object')

In [22]:
images[images['media_type'] == 'image/jpeg'].head()

archive.host_country                                      archive.notes  \
2                  USA  Images available via API at https://catalog.ar...   
3                  USA  Images available via API at https://catalog.ar...   
4                  USA  Images available via API at https://catalog.ar...   
5                  USA  Images available via API at https://catalog.ar...   
6                  USA  Images available via API at https://catalog.ar...   

  document.accession_to_archive_date document.contact_person  \
2                         2016-08-19              Kevin Wood   
3                         2016-08-19              Kevin Wood   
4                         2016-08-19              Kevin Wood   
5                         2016-08-19              Kevin Wood   
6                         2016-08-19              Kevin Wood   

   document.id_within_archive document.id_within_archive_type document.notes  \
2                    38547962                         NARA ID                  
3                    38547962                         NARA ID                  
4                    38547962                         NARA ID                  
5                    38547962                         NARA ID                  
6                    38547962                         NARA ID                  

  document.record_type document.rights_statement  \
2          ships' logs         CC0 Public Domain   
3          ships' logs         CC0 Public Domain   
4          ships' logs         CC0 Public Domain   
5          ships' logs         CC0 Public Domain   
6          ships' logs         CC0 Public Domain   

  document.standardized_region_list document.start_date  \
2                    north_atlantic          1957-09-30   
3                    north_atlantic          1957-09-30   
4                    north_atlantic          1957-09-30   
5                    north_atlantic          1957-09-30   
6                    north_atlantic          1957-09-30   

                                           file_path  media_type  \
2  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   
3  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   
4  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   
5  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   
6  2019-12-16/data/storis-wmec-38-1957-logbooks_0...  image/jpeg   

  platform.name                              uuid  
2   USCG Storis  37ea9580210611eaba89ac1f6bab1d82  
3   USCG Storis  3802945c210611eaba89ac1f6bab1d82  
4   USCG Storis  38198212210611eaba89ac1f6bab1d82  
5   USCG Storis  38308494210611eaba89ac1f6bab1d82  
6   USCG Storis  3847d224210611eaba89ac1f6bab1d82

Trying to subset a Series without any idea of what's good.

In [23]:
images.shape

(22, 15)

In [24]:
# explore the countries data
# rows (or entries)
countries.sample()

country google_country_code country_group   name_en    name_fr   name_de  \
24      sk                  SK            eu  Slovakia  Slovaquie  Slowakei   

     latitude  longitude  
24  48.672644  19.700032

In [25]:
# columns (or fields)
countries.columns

Index(['country', 'google_country_code', 'country_group', 'name_en', 'name_fr',
       'name_de', 'latitude', 'longitude'],
      dtype='object')

In [26]:
countries.shape

(30, 8)

In [27]:
countries.latitude 

0     47.696554
1     50.501045
2     42.725674
3     44.746643
4     35.129141
5     49.803531
6     55.939684
7     58.592469
8     64.950159
9     46.710994
10    51.163825
11    39.698467
12    47.161163
13    53.415260
14    42.504191
15    56.880117
16    55.173687
17    49.815319
18    35.902422
19    52.108118
20    64.556460
21    51.918907
22    39.558069
23    45.942611
24    48.672644
25    46.149259
26    39.895013
27    62.198468
28    38.952942
29    54.315447
Name: latitude, dtype: float64

In [28]:
countries.longitude

0     13.345980
1      4.476674
2     25.482322
3     15.340844
4     33.428682
5     15.474998
6      9.516689
7     25.806950
8     26.067564
9      1.718561
10    10.454048
11    21.577256
12    19.504265
13    -8.239122
14    12.573787
15    24.606555
16    23.943168
17     6.133352
18    14.447461
19     5.330198
20    12.665766
21    19.134334
22    -7.844941
23    24.990152
24    19.700032
25    14.986617
26    -2.988296
27    14.896307
28    35.439795
29    -2.232612
Name: longitude, dtype: float64

In [29]:
# concatenate along the first axis
pd.concat([countries['latitude'],countries['longitude']])

0     47.696554
1     50.501045
2     42.725674
3     44.746643
4     35.129141
5     49.803531
6     55.939684
7     58.592469
8     64.950159
9     46.710994
10    51.163825
11    39.698467
12    47.161163
13    53.415260
14    42.504191
15    56.880117
16    55.173687
17    49.815319
18    35.902422
19    52.108118
20    64.556460
21    51.918907
22    39.558069
23    45.942611
24    48.672644
25    46.149259
26    39.895013
27    62.198468
28    38.952942
29    54.315447
0     13.345980
1      4.476674
2     25.482322
3     15.340844
4     33.428682
5     15.474998
6      9.516689
7     25.806950
8     26.067564
9      1.718561
10    10.454048
11    21.577256
12    19.504265
13    -8.239122
14    12.573787
15    24.606555
16    23.943168
17     6.133352
18    14.447461
19     5.330198
20    12.665766
21    19.134334
22    -7.844941
23    24.990152
24    19.700032
25    14.986617
26    -2.988296
27    14.896307
28    35.439795
29    -2.232612
dtype: float64

In [30]:
# concatenate along the second axis
pd.concat([countries['latitude'],countries['longitude']], axis=1)

latitude  longitude
0   47.696554  13.345980
1   50.501045   4.476674
2   42.725674  25.482322
3   44.746643  15.340844
4   35.129141  33.428682
5   49.803531  15.474998
6   55.939684   9.516689
7   58.592469  25.806950
8   64.950159  26.067564
9   46.710994   1.718561
10  51.163825  10.454048
11  39.698467  21.577256
12  47.161163  19.504265
13  53.415260  -8.239122
14  42.504191  12.573787
15  56.880117  24.606555
16  55.173687  23.943168
17  49.815319   6.133352
18  35.902422  14.447461
19  52.108118   5.330198
20  64.556460  12.665766
21  51.918907  19.134334
22  39.558069  -7.844941
23  45.942611  24.990152
24  48.672644  19.700032
25  46.149259  14.986617
26  39.895013  -2.988296
27  62.198468  14.896307
28  38.952942  35.439795
29  54.315447  -2.232612

## 2. Rename, Index, and Slice <a id="section2"/>
Back to the entire unemployment data set. You may have noticed that the `month` column also includes the year. Let's go ahead and rename it.

In [31]:
pd.DataFrame.rename?

In [161]:
unemployment = pd.read_csv('data/country_total.csv')
unemployment.columns

Index(['country', 'seasonality', 'month', 'unemployment', 'unemployment_rate'], dtype='object')

In [162]:
unemployment.rename(columns={'month' : 'year_month'}, inplace=True)

In [163]:
unemployment.columns

Index(['country', 'seasonality', 'year_month', 'unemployment',
       'unemployment_rate'],
      dtype='object')

In [164]:
unemployment.isna().any()

country              False
seasonality          False
year_month           False
unemployment         False
unemployment_rate     True
dtype: bool

In [165]:
images.columns

Index(['archive.host_country', 'archive.notes',
       'document.accession_to_archive_date', 'document.contact_person',
       'document.id_within_archive', 'document.id_within_archive_type',
       'document.notes', 'document.record_type', 'document.rights_statement',
       'document.standardized_region_list', 'document.start_date',
       'image.file_path', 'image.media_type', 'platform.name', 'image.uuid'],
      dtype='object')

In [166]:
images.rename(columns={'uuid' : 'image.uuid', 'file_path' : 'image.file_path', 'media_type' : 'image.media_type'}, inplace=True)

In [167]:
images.rename(str.upper, axis='columns').head()
#rename the columns via a mapping

ARCHIVE.HOST_COUNTRY                                      ARCHIVE.NOTES  \
0                  USA  Images available via API at https://catalog.ar...   
1                  USA  Images available via API at https://catalog.ar...   
2                  USA  Images available via API at https://catalog.ar...   
3                  USA  Images available via API at https://catalog.ar...   
4                  USA  Images available via API at https://catalog.ar...   

  DOCUMENT.ACCESSION_TO_ARCHIVE_DATE DOCUMENT.CONTACT_PERSON  \
0                         2016-08-19              Kevin Wood   
1                         2016-08-19              Kevin Wood   
2                         2016-08-19              Kevin Wood   
3                         2016-08-19              Kevin Wood   
4                         2016-08-19              Kevin Wood   

   DOCUMENT.ID_WITHIN_ARCHIVE DOCUMENT.ID_WITHIN_ARCHIVE_TYPE DOCUMENT.NOTES  \
0                    38547962                         NARA ID                  
1                    38547962                         NARA ID                  
2                    38547962                         NARA ID                  
3                    38547962                         NARA ID                  
4                    38547962                         NARA ID                  

  DOCUMENT.RECORD_TYPE DOCUMENT.RIGHTS_STATEMENT  \
0          ships' logs         CC0 Public Domain   
1          ships' logs         CC0 Public Domain   
2          ships' logs         CC0 Public Domain   
3          ships' logs         CC0 Public Domain   
4          ships' logs         CC0 Public Domain   

  DOCUMENT.STANDARDIZED_REGION_LIST DOCUMENT.START_DATE  \
0                    north_atlantic          1957-09-30   
1                    north_atlantic          1957-09-30   
2                    north_atlantic          1957-09-30   
3                    north_atlantic          1957-09-30   
4                    north_atlantic          1957-09-30   

                                     IMAGE.FILE_PATH IMAGE.MEDIA_TYPE  \
0                       2019-12-16/data/metadata.csv       text/plain   
1  2019-12-16/data/nara_id_storis-wmec-38-1957-lo...       text/plain   
2  2019-12-16/data/storis-wmec-38-1957-logbooks_0...       image/jpeg   
3  2019-12-16/data/storis-wmec-38-1957-logbooks_0...       image/jpeg   
4  2019-12-16/data/storis-wmec-38-1957-logbooks_0...       image/jpeg   

  PLATFORM.NAME                        IMAGE.UUID  
0   USCG Storis                               NaN  
1   USCG Storis                               NaN  
2   USCG Storis  37ea9580210611eaba89ac1f6bab1d82  
3   USCG Storis  3802945c210611eaba89ac1f6bab1d82  
4   USCG Storis  38198212210611eaba89ac1f6bab1d82

In [168]:
images.rename(lambda x:2*x, axis='index').head()
#rename the indices

archive.host_country                                      archive.notes  \
0                  USA  Images available via API at https://catalog.ar...   
2                  USA  Images available via API at https://catalog.ar...   
4                  USA  Images available via API at https://catalog.ar...   
6                  USA  Images available via API at https://catalog.ar...   
8                  USA  Images available via API at https://catalog.ar...   

  document.accession_to_archive_date document.contact_person  \
0                         2016-08-19              Kevin Wood   
2                         2016-08-19              Kevin Wood   
4                         2016-08-19              Kevin Wood   
6                         2016-08-19              Kevin Wood   
8                         2016-08-19              Kevin Wood   

   document.id_within_archive document.id_within_archive_type document.notes  \
0                    38547962                         NARA ID                  
2                    38547962                         NARA ID                  
4                    38547962                         NARA ID                  
6                    38547962                         NARA ID                  
8                    38547962                         NARA ID                  

  document.record_type document.rights_statement  \
0          ships' logs         CC0 Public Domain   
2          ships' logs         CC0 Public Domain   
4          ships' logs         CC0 Public Domain   
6          ships' logs         CC0 Public Domain   
8          ships' logs         CC0 Public Domain   

  document.standardized_region_list document.start_date  \
0                    north_atlantic          1957-09-30   
2                    north_atlantic          1957-09-30   
4                    north_atlantic          1957-09-30   
6                    north_atlantic          1957-09-30   
8                    north_atlantic          1957-09-30   

                                     image.file_path image.media_type  \
0                       2019-12-16/data/metadata.csv       text/plain   
2  2019-12-16/data/nara_id_storis-wmec-38-1957-lo...       text/plain   
4  2019-12-16/data/storis-wmec-38-1957-logbooks_0...       image/jpeg   
6  2019-12-16/data/storis-wmec-38-1957-logbooks_0...       image/jpeg   
8  2019-12-16/data/storis-wmec-38-1957-logbooks_0...       image/jpeg   

  platform.name                        image.uuid  
0   USCG Storis                               NaN  
2   USCG Storis                               NaN  
4   USCG Storis  37ea9580210611eaba89ac1f6bab1d82  
6   USCG Storis  3802945c210611eaba89ac1f6bab1d82  
8   USCG Storis  38198212210611eaba89ac1f6bab1d82

The `.rename()` method allows you to modify index labels and/or column names. As you can see, we passed a `dict` to the `columns` parameter, with the original name as the key and the new name as the value. Importantly, we also set the `inplace` parameter to `True`, which modifies the *actual* `DataFrame`, not a copy of it.

It might also make sense to separate the data in `year_month` into two separate columns. To do this, you'll need to know how to select a single column. We can either use bracket (`[]`) or dot notation (referred to as *attribute access*).

In [169]:
unemployment['year_month'].head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

In [170]:
images['image.uuid'].head()

0                                 NaN
1                                 NaN
2    37ea9580210611eaba89ac1f6bab1d82
3    3802945c210611eaba89ac1f6bab1d82
4    38198212210611eaba89ac1f6bab1d82
Name: image.uuid, dtype: object

It is preferrable to use the bracket notation as a column name might inadvertently have the same name as a `DataFrame` (or `Series`) method. In addition, only bracket notation can be used to create a new column. If you try and use attribute access to create a new column, you'll create a new attribute, *not* a new column.

When selecting a single column, we have a `pandas` `Series` object, which is a single vector of data (e.g., a NumPy array) with "an associated array of data labels, called its *index*." A `DataFrame` also has an index. In our example, the indices are an array of sequential integers, which is the default. You can find them in the left-most position, without a column label.

Indices need not be a sequence of integers. They can, for example, be dates or strings. Note that indices do *not* need to be unique.

Indices, like column names, can be used to select data. Indices can be used to select particular rows. In fact, you can do something like `.head()` with slicing using the `[]` operator.

In [177]:
#np.random.rand?

In [178]:
unemployment['random'] = np.random.rand(20796)
unemployment.head()
# append a dim 20796 Series of random floats distributed uniformly in [0,1)

country seasonality  year_month  unemployment  unemployment_rate    random
0      at         nsa     1993.01        171000                4.5  0.393842
1      at         nsa     1993.02        175000                4.6  0.459397
2      at         nsa     1993.03        166000                4.4  0.529578
3      at         nsa     1993.04        157000                4.1  0.817504
4      at         nsa     1993.05        147000                3.9  0.025245

In [179]:
unemployment.drop(columns=['random'], inplace=True)
# drop that Series

Before we continue, let's look at a few useful ways to index data&mdash;that is, select rows.

`.loc` primarily works with string labels. It accepts a single label, a list (or array) of labels, or a slice of labels (e.g., `'a' : 'f'`).

In [180]:
unemployment[4:5]

country seasonality  year_month  unemployment  unemployment_rate
4      at         nsa     1993.05        147000                3.9

In [181]:
pd.DataFrame(unemployment.loc[4])

4
country                 at
seasonality            nsa
year_month         1993.05
unemployment        147000
unemployment_rate      3.9

In [182]:
type(unemployment.loc[4:5]), type(unemployment.loc[4])
# yet different indexing returns distinct python objects

(pandas.core.frame.DataFrame, pandas.core.series.Series)

Let's create a `DataFrame` to see how this works. (This is based on an [example](https://github.com/fonnesbeck/scipy2015_tutorial/blob/master/notebooks/1.%20Data%20Preparation.ipynb) from Chris Fonnesbeck's [Computational Statistics II Tutorial](https://github.com/fonnesbeck/scipy2015_tutorial).)

In [183]:
bacteria = pd.DataFrame({'bacteria_counts' : [632, 1638, 569, 115],
                         'other_feature' : [438, 833, 234, 298]},
                         index=['Firmicutes', 'Proteobacteria', 'Actinobacteria', 'Bacteroidetes'])
bacteria

bacteria_counts  other_feature
Firmicutes                  632            438
Proteobacteria             1638            833
Actinobacteria              569            234
Bacteroidetes               115            298

Notice that we pass in a `dict`, where the keys correspond to column names and the values to the data. In this example, we've also set the indices&mdash;strings in this case&mdash;to be the taxon of each bacterium.

Now, if we're interested in the values (row) associated with "Actinobacteria," we can use `.loc` and the index name.

In [184]:
bacteria.loc['Actinobacteria']

bacteria_counts    569
other_feature      234
Name: Actinobacteria, dtype: int64

This returns the column values for the specified row. Interestingly, we could have also used "positional indexing," even though the indices are strings.

In [185]:
bacteria.iloc[2]

bacteria_counts    569
other_feature      234
Name: Actinobacteria, dtype: int64

In [186]:
bacteria[2:3]

bacteria_counts  other_feature
Actinobacteria              569            234

Again, accessing `Firmicutes` by name, then by positional index.

In [187]:
bacteria.loc['Firmicutes']

bacteria_counts    632
other_feature      438
Name: Firmicutes, dtype: int64

In [188]:
bacteria.iloc[0]

bacteria_counts    632
other_feature      438
Name: Firmicutes, dtype: int64

In [189]:
type(bacteria.loc['Firmicutes'])

pandas.core.series.Series

In [190]:
#check that it's legit
bacteria.index[0] == 'Firmicutes'

True

Again, we can use positional slicing.

In [191]:
bacteria[:1]

bacteria_counts  other_feature
Firmicutes              632            438

In [192]:
bacteria[:1][:1] #it's not a matrix!

bacteria_counts  other_feature
Firmicutes              632            438

In [193]:
type(bacteria[:1])

pandas.core.frame.DataFrame

The difference is that the former returns a `Series` because we selected a single label, while the latter returns a `DataFrame` because we selected a range of positions.

Let's return to our unemployment data. Another indexing option, `.iloc`, primarily works with integer positions. To select specific rows, we can do the following.

In [194]:
#pd.DataFrame.iloc?

In [195]:
unemployment.iloc[[1, 5, 6, 9]]

country seasonality  year_month  unemployment  unemployment_rate
1      at         nsa     1993.02        175000                4.6
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
9      at         nsa     1993.10        141000                3.7

In [196]:
unemployment.iloc[[9,6,5,1]]

country seasonality  year_month  unemployment  unemployment_rate
9      at         nsa     1993.10        141000                3.7
6      at         nsa     1993.07        128000                3.4
5      at         nsa     1993.06        134000                3.5
1      at         nsa     1993.02        175000                4.6

In [197]:
unemployment.iloc[range(20)]

country seasonality  year_month  unemployment  unemployment_rate
0       at         nsa     1993.01        171000                4.5
1       at         nsa     1993.02        175000                4.6
2       at         nsa     1993.03        166000                4.4
3       at         nsa     1993.04        157000                4.1
4       at         nsa     1993.05        147000                3.9
5       at         nsa     1993.06        134000                3.5
6       at         nsa     1993.07        128000                3.4
7       at         nsa     1993.08        130000                3.4
8       at         nsa     1993.09        132000                3.5
9       at         nsa     1993.10        141000                3.7
10      at         nsa     1993.11        156000                4.1
11      at         nsa     1993.12        169000                4.4
12      at         nsa     1994.01        180000                4.7
13      at         nsa     1994.02        180000                4.7
14      at         nsa     1994.03        164000                4.3
15      at         nsa     1994.04        152000                4.0
16      at         nsa     1994.05        143000                3.8
17      at         nsa     1994.06        127000                3.4
18      at         nsa     1994.07        120000                3.2
19      at         nsa     1994.08        121000                3.2

We can select a range of rows and specify the step value.

In [198]:
unemployment.iloc[25:50:5]

country seasonality  year_month  unemployment  unemployment_rate
25      at         nsa     1995.02        174000                4.5
30      at         nsa     1995.07        123000                3.3
35      at         nsa     1995.12        175000                4.7
40      at         nsa     1996.05        159000                4.3
45      at         nsa     1996.10        146000                3.9

(Note: As is typical in Python, the end position is not included. Therefore, we don't see the row associated with the index 50.)

Indexing is important. You'll use it a lot. Below, we'll show how to index based on data values.



### Challenge 4: Renaming `bacteria`

The "other_feature" column in our `bacteria` table isn't very descriptive. Suppose we know that "other_feature" refers to a second set of bacteria count observations. Use the `rename` method to give "other_feature" a more descriptive name.

In [199]:
# rename "other_feature" in bacteria
bacteria.rename({'other_feature':'bacteria_recounts'}, axis='columns', inplace=True)
bacteria

bacteria_counts  bacteria_recounts
Firmicutes                  632                438
Proteobacteria             1638                833
Actinobacteria              569                234
Bacteroidetes               115                298

### Challenge 5: Indexing to get a specific value

Both `loc` and `iloc` can be used to select a particular value if they are given two arguments. The first argument is the name (when using `loc`) or index number (when using `iloc`) of the *row* you want, while the second argument is the name or index number of the *column* you want.

Using `loc`, select "Bacteroidetes" and "bacteria_counts" to get the count of Bacteroidetes.

BONUS: how could you do the same task using `iloc`?

In [200]:
bacteria.loc['Bacteroidetes', 'bacteria_counts'], type(bacteria.loc['Bacteroidetes', 'bacteria_counts'])

(115, numpy.int64)

In [201]:
bacteria.iloc[3, 0], type(bacteria.iloc[3, 0])

(115, numpy.int64)

### Challenge 6: Indexing multiple rows and columns

Both `loc` and `iloc` can be used to select subsets of columns *and* rows at the same time if they are given lists (and/or slices, for `iloc`] as their two arguments. 

Using `iloc` on the `unemployment` DataFrame, get:
* every row starting at row 4 and ending at row 7
* the 0th, 2nd, and 3rd columns

BONUS: how could you do the same task using `loc`?

In [202]:
unemployment.iloc[4:8:1]

country seasonality  year_month  unemployment  unemployment_rate
4      at         nsa     1993.05        147000                3.9
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
7      at         nsa     1993.08        130000                3.4

In [203]:
unemployment.iloc[[0,2,3]]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1

In [204]:
unemployment.loc[4:7]

country seasonality  year_month  unemployment  unemployment_rate
4      at         nsa     1993.05        147000                3.9
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
7      at         nsa     1993.08        130000                3.4

In [205]:
unemployment.loc[[0,2,3]]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1

## 3. Manipulating Columns: Renaming, Creating, Reordering

In [206]:
unemployment.sample()


country seasonality  year_month  unemployment  unemployment_rate
14108      nl          sa     1995.06        527000                7.1

So, we still want to **split `year_month` into two separate columns.** Above, we saw that this column is type (technically, `dtype`) `float64`. We'll first extract the year using the `.astype()` method. This allows for type casting&mdash;basically converting from one type to another. We'll then subtract this value from `year_month`&mdash;to get the decimal portion of the value&mdash;and multiply the result by 100 and convert to `int`.

For more information on `pandas` `dtype`s, check the documentation [here](http://pandas.pydata.org/pandas-docs/stable/basics.html#dtypes).

Just to prove a point, here's a mistake I made when downcasting the year `1993.0` to `int8`.

In [278]:
def downcasttoint(n, k):
    return ((n + 2**(k-1)) % 2**k) - 2**(k-1)

In [294]:
d = np.linspace(1993, 1994, 11)
d

array([1993. , 1993.1, 1993.2, 1993.3, 1993.4, 1993.5, 1993.6, 1993.7,
       1993.8, 1993.9, 1994. ])

In [299]:
downcasttoint(d, 8) # get unique representatives of the series in the interval [-128, 127)

array([-55. , -54.9, -54.8, -54.7, -54.6, -54.5, -54.4, -54.3, -54.2,
       -54.1, -54. ])

In [304]:
# numpy downcasts to integers by truncation, not by flooring, e.g.,
# 1993.0 to -55.0 to -55 
# 1993.1 to -54.9 to -54
downcasttoint(d, 8).astype(np.int8) 

array([-55, -54, -54, -54, -54, -54, -54, -54, -54, -54, -54], dtype=int8)

~~`2^11`~~ (avoid `^`!) `2**11` radix is just too darn small to handle year numbers larger than `1024 - 1`!

In [314]:
d = pd.Series(np.arange(1993, 1994, 0.1))
d2 = downcasttoint(d, 11)
d3 = d2.astype(int)
pd.concat([d, d2, d3], axis=1)

0     1   2
0  1993.0 -55.0 -55
1  1993.1 -54.9 -54
2  1993.2 -54.8 -54
3  1993.3 -54.7 -54
4  1993.4 -54.6 -54
5  1993.5 -54.5 -54
6  1993.6 -54.4 -54
7  1993.7 -54.3 -54
8  1993.8 -54.2 -54
9  1993.9 -54.1 -54

In [219]:
pd.DataFrame.astype?

In [218]:
np.floor?

In [315]:
unemployment['year'] = np.floor(unemployment['year_month']).astype(np.int16)
unemployment.head()

country seasonality  year_month  unemployment  unemployment_rate  year
0      at         nsa     1993.01        171000                4.5  1993
1      at         nsa     1993.02        175000                4.6  1993
2      at         nsa     1993.03        166000                4.4  1993
3      at         nsa     1993.04        157000                4.1  1993
4      at         nsa     1993.05        147000                3.9  1993

In [316]:
country_names.head()

country country_group   name_en
0      at            eu   Austria
1      be            eu   Belgium
2      bg            eu  Bulgaria
3      hr        non-eu   Croatia
4      cy            eu    Cyprus

In this case, we're casting the floating point values to integers. In Python, this [truncates the decimals](https://docs.python.org/2/library/stdtypes.html#numeric-types-int-float-long-complex).

Finally, let's create our **month** variable as described above. (Because of the truncating that occurs when casting to `int`, we first round the values to the nearest whole number.)

In [318]:
unemployment['month'] = ((unemployment['year_month'] - unemployment['year']) * 100).round(0).astype(np.int8)
#int8 is safe to hold 1 through 12

In [319]:
type(unemployment.month[0]), type(unemployment.year[0])

(numpy.int8, numpy.int16)

In [320]:
unemployment.head()

country seasonality  year_month  unemployment  unemployment_rate  year  \
0      at         nsa     1993.01        171000                4.5  1993   
1      at         nsa     1993.02        175000                4.6  1993   
2      at         nsa     1993.03        166000                4.4  1993   
3      at         nsa     1993.04        157000                4.1  1993   
4      at         nsa     1993.05        147000                3.9  1993   

   month  
0      1  
1      2  
2      3  
3      4  
4      5

To create the `month` column, we subtracted two vectors. This resulted in the decimal value in `year_month`. To transform the values to integers, we multiplied by 100.

Now, let's say we wanted to **reorder the columns** in the `DataFrame`. For this, we use bracket notation again, passing in a list of column names in the order we'd like to see them.

In [321]:
unemployment = unemployment[['country', 'seasonality','year', 'month','year_month','unemployment', 'unemployment_rate']]
# make evil state changes to unemployment

In [322]:
unemployment.head()

country seasonality  year  month  year_month  unemployment  \
0      at         nsa  1993      1     1993.01        171000   
1      at         nsa  1993      2     1993.02        175000   
2      at         nsa  1993      3     1993.03        166000   
3      at         nsa  1993      4     1993.04        157000   
4      at         nsa  1993      5     1993.05        147000   

   unemployment_rate  
0                4.5  
1                4.6  
2                4.4  
3                4.1  
4                3.9

### Challenge 7: Another way to get the month

If you didn't know that casting floats to ints truncates the decimals in Python, you could have used NumPy's `floor()` function. `np.floor` takes an array or Pandas Series of floats as its argument, and returns an array or Series where every float has been rounded down to the nearest whole number. 


Use `np.floor` to round the values in the "year_month" column down so we can cast them as integer years. Note that the types are still floats, so we'll still need to use `astype` to typecast.

In [323]:
import numpy as np

In [324]:
# select the "year_month" column
year_month = unemployment['year_month']

# use np.floor on year_month to get the years as floats
years_by_floor = np.floor(year_month)

# cast years_by_floor to integers using astype(int)
int_years = years_by_floor.astype(int)

# check that this gets the same answers as our first approach
# this should return True
(unemployment['year_month'].astype(int) == int_years).all()

True

In [325]:
#np.all?

The last line of code in the previous cell does an element-wise comparison of the values in the corresponding arrays. The `.all()` method checks whether *all* elements are `True`.

In [326]:
unemployment.drop(columns=['year_month'], inplace=True)
unemployment.head()

country seasonality  year  month  unemployment  unemployment_rate
0      at         nsa  1993      1        171000                4.5
1      at         nsa  1993      2        175000                4.6
2      at         nsa  1993      3        166000                4.4
3      at         nsa  1993      4        157000                4.1
4      at         nsa  1993      5        147000                3.9

### combining datasets: concat and append

In [327]:
def make_df(cols, ind):
    """Quickly make a DataFrame"""
    data = {c : [str(c) + str(i) for i in ind] 
            for c in cols}
    return pd.DataFrame(data, ind)
make_df('ABC', range(3))

A   B   C
0  A0  B0  C0
1  A1  B1  C1
2  A2  B2  C2

In [328]:
x = np.arange(0, 8, 2).reshape(2,2)
np.concatenate([x,x], axis=1)

array([[0, 2, 0, 2],
       [4, 6, 4, 6]])

In [329]:
ser1 = pd.Series(['A', 'B', 'C'], index = range(1,4))
ser2 = pd.Series(['D', 'E', 'F'], index = range(4,7))
pd.concat([ser1, ser2])

1    A
2    B
3    C
4    D
5    E
6    F
dtype: object

In [330]:
df1 = make_df('AB', [1,2])
df2 = make_df('AB', [3,4])
pd.concat([df1, df2], axis=0) #default

A   B
1  A1  B1
2  A2  B2
3  A3  B3
4  A4  B4

In [331]:
pd.concat([df1, df2], axis=1) #outer merge

A    B    A    B
1   A1   B1  NaN  NaN
2   A2   B2  NaN  NaN
3  NaN  NaN   A3   B3
4  NaN  NaN   A4   B4

In [332]:
d1 = make_df('AB', [0,1])
d2 = make_df('CD', [0,1])
# we make a dataframe with duplicate indices
pd.concat([d1, d2], join='outer', sort=True)

A    B    C    D
0   A0   B0  NaN  NaN
1   A1   B1  NaN  NaN
0  NaN  NaN   C0   D0
1  NaN  NaN   C1   D1

In [333]:
try:
    pd.concat([d1, d2], verify_integrity=True)
except ValueError as e:
    print("Value Error! " + str(e))

Value Error! Indexes have overlapping values: Int64Index([0, 1], dtype='int64')


/home/colton/.local/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


In [334]:
pd.concat([d1, d2], ignore_index=True, sort=False)

A    B    C    D
0   A0   B0  NaN  NaN
1   A1   B1  NaN  NaN
2  NaN  NaN   C0   D0
3  NaN  NaN   C1   D1

In [335]:
pd.concat([d1, d2], axis=1) # "horizontal" merge, ie., along column axis

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1

## 4. Merging

So far, our `DataFrame` is organized in a reasonable way. But, we know we can do better. We're eventually going to be interested in the unemployment rate for each country. The trouble is, we don't exactly know what the values in `country` refer to. We can fix that by getting country names. These can be found in `countries.csv`. Run the following cell to import the countries data using a URL as the file path.

In [336]:
countries = pd.read_csv('data/countries.csv')

This file has lots of useful information. It even has the country names is three different languages.

Because the data we need is stored in two separate files, we'll want to merge the data somehow. Let's determine which column we can use to join this data. `country` looks like a good option. However, we don't need all of the columns in the `countries` `DataFrame`. To select certain columns, we use the name bracket notation we used to reorder the columns.

In [337]:
country_names = countries[['country', 'country_group', 'name_en']]

`pandas` includes an easy-to-use merge function. Let's use it to **merge the two `DataFrame`s on country code.**

In [338]:
unemployment.head()

country seasonality  year  month  unemployment  unemployment_rate
0      at         nsa  1993      1        171000                4.5
1      at         nsa  1993      2        175000                4.6
2      at         nsa  1993      3        166000                4.4
3      at         nsa  1993      4        157000                4.1
4      at         nsa  1993      5        147000                3.9

In [339]:
country_names.head()

country country_group   name_en
0      at            eu   Austria
1      be            eu   Belgium
2      bg            eu  Bulgaria
3      hr        non-eu   Croatia
4      cy            eu    Cyprus

In [340]:
u1 = pd.merge(unemployment, country_names, on='country')

In [341]:
u1.head()

country seasonality  year  month  unemployment  unemployment_rate  \
0      at         nsa  1993      1        171000                4.5   
1      at         nsa  1993      2        175000                4.6   
2      at         nsa  1993      3        166000                4.4   
3      at         nsa  1993      4        157000                4.1   
4      at         nsa  1993      5        147000                3.9   

  country_group  name_en  
0            eu  Austria  
1            eu  Austria  
2            eu  Austria  
3            eu  Austria  
4            eu  Austria

Merging is often more complex than this example. If you want to merge on multiple columns, you can pass a list of column names to the `on` parameter.

```
pd.merge(first, second, on=['name', 'id'])
```

For more information on merging, check the [documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging).

`pandas` also provides a `.merge()` method that can act on a `DataFrame`. You can read more about that [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html).

### Challenge 8: 

You may sometimes need to merge on columns with different names. To do so, use the `left_on` and `right_on` parameters, where the first listed `DataFrame` is the "left" one and the second is the "right." It might look something this.

```
pd.merge(one, two, left_on='city', right_on='city_name')
```

Suppose wanted to merge `unemployment` with a new DataFrame called `country_codes`, where the abbreviation for each country is in the column "c_code":

In [342]:
countries.head()

country google_country_code country_group   name_en   name_fr     name_de  \
0      at                  AT            eu   Austria  Autriche  Österreich   
1      be                  BE            eu   Belgium  Belgique     Belgien   
2      bg                  BG            eu  Bulgaria  Bulgarie   Bulgarien   
3      hr                  HR        non-eu   Croatia   Croatie    Kroatien   
4      cy                  CY            eu    Cyprus    Chypre      Zypern   

    latitude  longitude  
0  47.696554  13.345980  
1  50.501045   4.476674  
2  42.725674  25.482322  
3  44.746643  15.340844  
4  35.129141  33.428682

In [343]:
country_codes = countries.loc[:, ['country', 'name_de']].copy()

In [344]:
country_codes.rename({"country":"c_code"}, axis=1, inplace=True)
country_codes.head()

c_code     name_de
0     at  Österreich
1     be     Belgien
2     bg   Bulgarien
3     hr    Kroatien
4     cy      Zypern

In [345]:
u2 = pd.merge(unemployment, country_codes, left_on='country', right_on='c_code')
u2.head()

country seasonality  year  month  unemployment  unemployment_rate c_code  \
0      at         nsa  1993      1        171000                4.5     at   
1      at         nsa  1993      2        175000                4.6     at   
2      at         nsa  1993      3        166000                4.4     at   
3      at         nsa  1993      4        157000                4.1     at   
4      at         nsa  1993      5        147000                3.9     at   

      name_de  
0  Österreich  
1  Österreich  
2  Österreich  
3  Österreich  
4  Österreich

In [346]:
# drop duplicate columns by transposition
# https://stackoverflow.com/questions/14984119/python-pandas-remove-duplicate-columns
u3 = u2.T.drop_duplicates().T
# long runtime!

In [347]:
u3.head()

country seasonality  year month unemployment unemployment_rate     name_de
0      at         nsa  1993     1       171000               4.5  Österreich
1      at         nsa  1993     2       175000               4.6  Österreich
2      at         nsa  1993     3       166000               4.4  Österreich
3      at         nsa  1993     4       157000               4.1  Österreich
4      at         nsa  1993     5       147000               3.9  Österreich

## 5. Calculating Unique and Missing Values

In [348]:
u1.head()

country seasonality  year  month  unemployment  unemployment_rate  \
0      at         nsa  1993      1        171000                4.5   
1      at         nsa  1993      2        175000                4.6   
2      at         nsa  1993      3        166000                4.4   
3      at         nsa  1993      4        157000                4.1   
4      at         nsa  1993      5        147000                3.9   

  country_group  name_en  
0            eu  Austria  
1            eu  Austria  
2            eu  Austria  
3            eu  Austria  
4            eu  Austria

That's better. We now know that the abbreviation "at" corresponds to Austria. We might be curious to check what countries we have data for. The `Series` object includes a `.unique()` method. We'll use this to check the countries. We can select the name either using bracket or dot notation. (While we suggested using brackets above, it *is* sometimes easier to use dot notation. Just be careful.)

In [349]:
u1['name_en'].unique()

array(['Austria', 'Belgium', 'Bulgaria', 'Cyprus', 'Czech Republic',
       'Germany (including  former GDR from 1991)', 'Denmark', 'Estonia',
       'Spain', 'Finland', 'France', 'Greece', 'Croatia', 'Hungary',
       'Ireland', 'Italy', 'Lithuania', 'Luxembourg', 'Latvia', 'Malta',
       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Sweden',
       'Slovenia', 'Slovakia', 'Turkey', 'United Kingdom'], dtype=object)

To get a count of the **number of unique countries,** we can either wrap the above code with `len()` to get the number of items in the array, or we can use the  `Series.nunique()` method.

In [350]:
u1['name_en'].nunique()

30

It might be more interesting to know **how many observations** we actually have. `pandas` has a `Series` method called `.value_counts()` that returns the counts for the unique values in the `Series`.

In [351]:
dnd = u3['name_de'].value_counts()

In [352]:
type(dnd), dnd.shape, dnd.index

(pandas.core.series.Series,
 (30,),
 Index(['Luxemburg', 'Frankreich', 'Irland', 'Spanien', 'Dänemark', 'Portugal',
        'Belgien', 'Niederlande', 'Schweden', 'Vereinigtes Königreich',
        'Italien', 'Finnland', 'Norwegen', 'Österreich', 'Bulgarien',
        'Slowakei', 'Malta', 'Ungarn', 'Polen', 'Slowenien',
        'Deutschland (einschließlich der ehemaligen DDR seit 1991)',
        'Tschechische Republik', 'Litauen', 'Lettland', 'Griechenland',
        'Rumänien', 'Zypern', 'Estland', 'Kroatien', 'Türkei'],
       dtype='object'))

In [353]:
u1['name_en'].value_counts()

Denmark                                      1008
Netherlands                                  1008
Luxembourg                                   1008
Portugal                                     1008
Ireland                                      1008
Belgium                                      1008
Spain                                        1008
Sweden                                       1008
France                                       1008
United Kingdom                               1002
Italy                                         924
Finland                                       828
Norway                                        786
Austria                                       648
Bulgaria                                      576
Poland                                        576
Malta                                         576
Slovakia                                      576
Slovenia                                      576
Hungary                                       576


By default, the `Series` is sorted by values. If you'd like it sorted by index&mdash;country name in this case&mdash;append the `.sort_index()` method.

In [354]:
pd.Series.sort_index?

In [357]:
u1['name_en'].value_counts().sort_index()

Austria                                       648
Belgium                                      1008
Bulgaria                                      576
Croatia                                       324
Cyprus                                        396
Czech Republic                                468
Denmark                                      1008
Estonia                                       387
Finland                                       828
France                                       1008
Germany (including  former GDR from 1991)     504
Greece                                        450
Hungary                                       576
Ireland                                      1008
Italy                                         924
Latvia                                        459
Lithuania                                     459
Luxembourg                                   1008
Malta                                         576
Netherlands                                  1008


This will be useful for our analysis. The maximum number of observations for a given country for this time period is 1,008 observations. We'll note that certain countries, such as Turkey, have far less data.

How about finding the **date range** for this data set? Let's look at the minimum and maximum years.

In [358]:
unemployment['year'].min(), unemployment['year'].max()

(1983, 2010)

Next, we should pause for a moment and think about what data we really care about. For our purposes, the variable of interest is `unemployment_rate`. The number of observations by country only reflect the number of instances of each country name in the dataset. It is possible, maybe even expected, to have some missing data within those instances. Let's find out **how many unemployment rate values are missing.**

In [359]:
unemployment['unemployment_rate'].isnull().sum()

945

But no year values are missing.

In [363]:
unemployment['year'].isnull().any()

False

The `.isnull()` method returns a corresponding boolean value for each entry in the unemployment rate `Series`. In Python `True` is equivalent to 1 and `False` is equivalent to 0. Thus, when we add the result (with `.sum()`), we get a count for the *total* number of missing values.

### Challenge 9: Exploring unemployment rates

What are the minimum and maximum unemployment rates in our data set? 

In [364]:
unemployment['unemployment_rate'].describe()

count    19851.000000
mean         8.179764
std          3.922533
min          1.100000
25%          5.200000
50%          7.600000
75%         10.000000
max         20.900000
Name: unemployment_rate, dtype: float64

Which unemployment rates are most and least common?

In [379]:
v = unemployment['unemployment_rate'].value_counts()
print(v[v == v.max()])
print(v[v == v.min()])

7.2    305
Name: unemployment_rate, dtype: int64
20.9    1
Name: unemployment_rate, dtype: int64


## 6. GroupBy


What if we'd like to know how many missing values exist at the *country* level? We can take the main part of what we had above and create a new column in the `DataFrame`.

In [388]:
u1['unemployment_rate_null'] = u1['unemployment_rate'].isnull()

To count the **number of missing values for each country,** we introduce the `.groupby()` method.

In [395]:
u1.groupby('name_en')['unemployment_rate_null'].sum()

name_en
Austria                                        0.0
Belgium                                        0.0
Bulgaria                                     180.0
Croatia                                      216.0
Cyprus                                         0.0
Czech Republic                                 0.0
Denmark                                        0.0
Estonia                                        0.0
Finland                                        0.0
France                                         0.0
Germany (including  former GDR from 1991)      0.0
Greece                                         0.0
Hungary                                       36.0
Ireland                                        0.0
Italy                                          0.0
Latvia                                         0.0
Lithuania                                      0.0
Luxembourg                                     0.0
Malta                                        180.0
Netherlands            

Let's explain what just happened. We start with our `DataFrame`. We tell `pandas` that we want to group the data by country name&mdash;that's what goes in the parentheses. Next, we need to tell it what column we'd like to perform the `.sum()` operation on. In this case, it's the indicator for whether or not the unemployment rate was missing.

As we saw above, the number of records for each country differs. We might, then, want to have the **missing values by country shown as percentages.** Let's create a new `DataFrame` for this.

We'll take the code from above and set the `as_index` parameter to `False`.

In [399]:
unemp_rate.groupby?

In [398]:
unemp_rate = u1.groupby('name_en', as_index=False)['unemployment_rate_null'].sum()

In [397]:
unemp_rate.head(3)

name_en  unemployment_rate_null
0   Austria                     0.0
1   Belgium                     0.0
2  Bulgaria                   180.0

`unemployment_rate` is a `DataFrame` with the information from cell 34, above. It's important to note that using `as_index=False` in `.groupby()` only works if the grouping column(s) are not the same as the columns on which we're performing the operation.

Also, to group by several columns, simply pass in a list of column names to `.groupby()`.

```
unemployment.groupby(['name_en', 'seasonality'])['unemployment_rate'].mean()
```

In [402]:
u1.groupby(['name_en', 'seasonality'])['unemployment_rate'].mean()

name_en         seasonality
Austria         nsa             4.278241
                sa              4.280556
                trend           4.276852
Belgium         nsa             8.475595
                sa              8.483036
                trend           8.483929
Bulgaria        nsa            11.649242
                sa             11.643939
                trend          11.640152
Croatia         nsa            11.727778
                sa                   NaN
                trend                NaN
Cyprus          nsa             4.607576
                sa              4.610606
                trend           4.606818
Czech Republic  nsa             7.233333
                sa              7.235256
                trend           7.237179
Denmark         nsa             6.013095
                sa              6.014583
                trend           6.008929
Estonia         nsa             9.981395
                sa              9.979070
                trend        

Now, let's add the number of observations by country to the `DataFrame`.

In [106]:
unemployment_rate['n_obs'] = unemployment.groupby('name_en')['name_en'].count().values

Here, we need to use the `values` attribute to get an array of the counts. Excluding `values` will result in a column full of `NaN`s. This is because the index in `unemployment.groupby('name_en')['name_en'].count()` is a list of the country names. When creating a new column, `pandas` tries to match on index. Recall that the default index values for a `DataFrame` is a sequence of integers.

Because we know (or have noticed) that the `.groupby()` function returns the values in alphabetical order, we can simply set the new column to the list of values, as we have done. You can, however, be more explicit and create another `DataFrame` and merge on country name.

Finally, let's create the column for the percentage of missing values.

In [107]:
unemployment_rate['null_percentage'] = unemployment_rate['unemployment_rate_null'] / unemployment_rate['n_obs']

In [108]:
unemployment_rate

name_en  unemployment_rate_null  n_obs  \
0                                     Austria                     0.0    648   
1                                     Belgium                     0.0   1008   
2                                    Bulgaria                   180.0    576   
3                                     Croatia                   216.0    324   
4                                      Cyprus                     0.0    396   
5                              Czech Republic                     0.0    468   
6                                     Denmark                     0.0   1008   
7                                     Estonia                     0.0    387   
8                                     Finland                     0.0    828   
9                                      France                     0.0   1008   
10  Germany (including  former GDR from 1991)                     0.0    504   
11                                     Greece                     0.0    450   
12                                    Hungary                    36.0    576   
13                                    Ireland                     0.0   1008   
14                                      Italy                     0.0    924   
15                                     Latvia                     0.0    459   
16                                  Lithuania                     0.0    459   
17                                 Luxembourg                     0.0   1008   
18                                      Malta                   180.0    576   
19                                Netherlands                     0.0   1008   
20                                     Norway                     0.0    786   
21                                     Poland                    72.0    576   
22                                   Portugal                     0.0   1008   
23                                    Romania                     0.0    423   
24                                   Slovakia                   108.0    576   
25                                   Slovenia                    36.0    576   
26                                      Spain                   117.0   1008   
27                                     Sweden                     0.0   1008   
28                                     Turkey                     0.0    210   
29                             United Kingdom                     0.0   1002   

    null_percentage  
0          0.000000  
1          0.000000  
2          0.312500  
3          0.666667  
4          0.000000  
5          0.000000  
6          0.000000  
7          0.000000  
8          0.000000  
9          0.000000  
10         0.000000  
11         0.000000  
12         0.062500  
13         0.000000  
14         0.000000  
15         0.000000  
16         0.000000  
17         0.000000  
18         0.312500  
19         0.000000  
20         0.000000  
21         0.125000  
22         0.000000  
23         0.000000  
24         0.187500  
25         0.062500  
26         0.116071  
27         0.000000  
28         0.000000  
29         0.000000

This is the second time we've called a `DataFrame` without something like `.head()`. Let's describe what it does. By default, `pandas` prints 60 rows. In this case, because there are only 30 countries, we see the entire `DataFrame`.

### Challenge 10: GroupBy 

Find the average unemployment rate for European Union vs. non-European Union countries. 

1. use `groupby` to group on "country_group"
2. select the "unemployment_rate" column
3. use `.mean()` to get the average

## 7. Exporting A DataFrame to csv <a id="section7"/>

As we can see, Croatia has lots of missing data. This `DataFrame` contains useful information&mdash;things to consider&mdash;when analyzing the data.

Suppose we wanted to save this as a .csv file. For this, we'd use the `.to_csv()` method.

In [109]:
unemployment_rate.to_csv('data/unemployment_missing.csv')

Let's look at our file.

In [110]:
!head -5 data/unemployment_missing.csv

,name_en,unemployment_rate_null,n_obs,null_percentage
0,Austria,0.0,648,0.0
1,Belgium,0.0,1008,0.0
2,Bulgaria,180.0,576,0.3125
3,Croatia,216.0,324,0.6666666666666666


By default, this method writes the indices. We probably don't want that. Let's edit the code. Let's also be explicit about the type of delimiter we're interested in. (Values can be separated by pipes (`|`), semicolons (`;`), tabs (`\t`), etc.)

In [111]:
unemployment_rate.to_csv('data/unemployment_missing.csv', index=False, sep=',')

In [112]:
!head -5 data/unemployment_missing.csv

name_en,unemployment_rate_null,n_obs,null_percentage
Austria,0.0,648,0.0
Belgium,0.0,1008,0.0
Bulgaria,180.0,576,0.3125
Croatia,216.0,324,0.6666666666666666


Much better!

Let's return to our main `DataFrame`. Now that we have the missing values information in `unemployment_rate`, we can **drop the last column** we added to `unemployment`.

In [113]:
unemployment.drop('unemployment_rate_null', axis=1, inplace=True)

It's important to specify the `axis` parameter. `axis=1` refers to columns (`axis=0` refers to rows.) The parameter `inplace=True` simply modifies the actual `DataFrame` rather than returning a new `DataFrame`.

In [114]:
unemployment.head()

country seasonality  year  month  unemployment  unemployment_rate  \
0      at         nsa   -55      1        171000                4.5   
1      at         nsa   -55      2        175000                4.6   
2      at         nsa   -55      3        166000                4.4   
3      at         nsa   -55      4        157000                4.1   
4      at         nsa   -55      5        147000                3.9   

  country_group  name_en  
0            eu  Austria  
1            eu  Austria  
2            eu  Austria  
3            eu  Austria  
4            eu  Austria

## 8. Dealing With Missing Values: Boolean Indexing <a id="section8"/>

Now that we know about the missing values, we have to deal with them. There are two main options:

* Fill the missing values with some other values.
* Do not use observations with missing values.
    * Depending on the analysis, we may want to exclude entire countries.
    
Because countries with missing unemployment rate data have at least 36 missing values, which is too many to fill, we'll take the second approach and **exclude missing values** from our primary analyses.

Instead of just getting rid of that data, it might make sense to store it in a separate `DataFrame`. This way, we could answer questions such as, "do missing values occur during certain months (or years) more frequently?" With this, we will introduce the concept of *boolean indexing* for filtering data.

In [115]:
unemployment_rate_missing = unemployment[unemployment['unemployment_rate'].isnull()]

Recall that `unemployment['unemployment_rate'].isnull()` produces an array of boolean values. We used this previously when counting the number of missing values, though we did not see its output. Let's see some of that now.

In [116]:
unemployment['unemployment_rate'].isnull()[:10]

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: unemployment_rate, dtype: bool

To create `unemployment_rate_missing`, we're indexing `unemployment` with the array above. This returns only the rows where the value in the array is `True`. Let's see if it worked.

In [117]:
unemployment_rate_missing.head()

country seasonality  year  month  unemployment  unemployment_rate  \
1656      bg         nsa   -53      1        391000                NaN   
1657      bg         nsa   -53      2        387000                NaN   
1658      bg         nsa   -53      3        378000                NaN   
1659      bg         nsa   -53      4        365000                NaN   
1660      bg         nsa   -53      5        346000                NaN   

     country_group   name_en  
1656            eu  Bulgaria  
1657            eu  Bulgaria  
1658            eu  Bulgaria  
1659            eu  Bulgaria  
1660            eu  Bulgaria

It is also possible to specify multiple conditions using the `&` operator, but each condition needs to be inside of parentheses. The `.isin()` method, which takes a `list` of values, is useful when you're interested in conditioning on multiple values on a given column. For example, if you want to select multiple countries.

Now, we're ready to remove the missing data in `unemployment`. To do this, we can use the `.dropna()` method.

In [118]:
unemployment.dropna(subset=['unemployment_rate'], inplace=True)

(Note that while we have dropped all observations for which `unemployment_rate == NaN`, this doesn't mean that all of our observations overlap exactly in time. We may find that there are dates where we have data for one country and no data for others.)

### Challenge 11: Boolean Indexing
Suppose we only want to look at unemployment data from the year 2000 or later. Use Boolean indexing to create a DataFrame with only these years.

1. Select the "year" column from `unemployment`
2. Using the year data, create a **mask**: an array of Booleans where each value is True if and only if the year is 2000 or later. Remember, you can use Boolean operators like `>`, `<`, and `==` on a column
3. Use the mask from step 2 to index `unemployment`

In [119]:
# select the year column from unemployment
year = ...

# create a mask
later_or_equal_2000 = ...

# Boolean index unemployment
...

Ellipsis

## 9. Sorting Values

At this point, you might be curious to know what the highest unemployment rates were. For this, we'll use the `DataFrame.sort_values()` method to **sort the data.**

In [120]:
unemployment.sort_values('unemployment_rate', ascending=False)[:5]

country seasonality  year  month  unemployment  unemployment_rate  \
15526      pl         nsa   -44      2       3531000               20.9   
15525      pl         nsa   -44      1       3520000               20.7   
15514      pl         nsa   -45      2       3460000               20.7   
5663       es          sa   -38      9       4773000               20.6   
15527      pl         nsa   -44      3       3475000               20.6   

      country_group name_en  
15526            eu  Poland  
15525            eu  Poland  
15514            eu  Poland  
5663             eu   Spain  
15527            eu  Poland

The above code creates a copy of the `DataFrame`, sorted in *descending* order, and prints the first five rows.

You may have noticed that the data set includes a `seasonality` column, which we haven't yet discussed. The unemployment rate in this data is actually calculated in three separate ways. Let's look at the values.

In [121]:
unemployment['seasonality'].unique()

array(['nsa', 'sa', 'trend'], dtype=object)

The three options above correspond to:

* not seasonally adjusted
* seasonally adjusted
* trend cycle

We'll stick with seasonally adjusted data so that the values are more comparable. Let's look at the highest unemployment rates in this context.

In [122]:
unemployment[unemployment['seasonality'] == 'sa'].sort_values('unemployment_rate', ascending=False)[:5]

country seasonality  year  month  unemployment  unemployment_rate  \
5664       es          sa   -38     10       4758000               20.6   
5663       es          sa   -38      9       4773000               20.6   
5662       es          sa   -38      8       4739000               20.5   
5665       es          sa   -38     11       4723000               20.4   
15702      pl          sa   -46     10       3471000               20.3   

      country_group name_en  
5664             eu   Spain  
5663             eu   Spain  
5662             eu   Spain  
5665             eu   Spain  
15702            eu  Poland

Spain has the highest seasonally adjusted unemployment rate.



## 10. Plotting With Pandas

The best way to get a sense of this data is to **plot it.** Next, we'll start to look at some basic plotting with `pandas`. Before we begin, let's sort the data by country and date. This is good practice and is especially important when using `pandas`'s `.plot()` method because the x-axis values are based on the indices. When we sort, the index values remain unchanged. Thus, we need to reset them. The `drop` parameter tells `pandas` to construct a `DataFrame` *without* adding a column.

In [123]:
unemployment.sort_values(['name_en', 'year_month'], inplace=True)
unemployment.reset_index(drop=True, inplace=True)

KeyError: 'year_month'

Let's take a look at Spain's unemployment rate (only because it was the highest) across time.

In [ ]:
spain = unemployment[(unemployment['name_en'] == 'Spain') &
                     (unemployment['seasonality'] == 'sa')]

In [ ]:
spain['unemployment_rate'].plot(figsize=(10, 8), color='#348ABD')

Note that the values along the x-axis represent the indices associated with Spain in the sorted `unemployment` `DataFrame`. Wouldn't it be nice if, instead, we could **show the time period** associated with the various unemployment rates for Spain? It might also be interesting to **compare** Spain's unemployment rate with its neighbor to the west, Portugal.

Let's first create a `DataFrame` that contains the unemployment data for both countries.

In [ ]:
ps = unemployment[(unemployment['name_en'].isin(['Portugal', 'Spain'])) &
                  (unemployment['seasonality'] == 'sa')]

Next, we'll **generate time series data** by converting our years and months into `datetime` objects. `pandas` provides a `to_datetime()` function that makes this relatively simple. It converts an argument&mdash;a single value or an array of values&mdash;to `datetime`. (Note that the return value [depends on the input](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html).) If we were interested in March 23, 1868, for example, we could do the following.

In [ ]:
pd.to_datetime('1868/3/23')

The argument doesn't necessarily have to be specified in the `yyyy/mm/dd` format. You could list it as `mm/dd/yyyy`, but it's a good idea to be explicit. As a result, we pass in a valid string format.

In [ ]:
pd.to_datetime('3/23/1868', format='%m/%d/%Y')

Let's create the `datetime` object and add it to the `DataFrame` as a column named `date`. For this, we'll use the `DataFrame.insert()` method.

In [ ]:
ps.insert(loc=0, column='date',
          value=pd.to_datetime(ps['year'].astype(str) + '/' + ps['month'].astype(str) + '/1'))

Finally, let's only keep certain columns, rename them, and reshape the `DataFrame`.

In [ ]:
ps = ps[['date', 'name_en', 'unemployment_rate']]
ps.columns = ['Time Period', 'Country', 'Unemployment Rate']
ps = ps.pivot(index='Time Period', columns='Country', values='Unemployment Rate')
ps.tail()

In [ ]:
ps.head()

Notice the indices.

In [ ]:
ps.plot(figsize=(10, 8), title='Unemployment Rate\n')

### Challenge 12: Plot without missing values
Note that there are some dates for which we lack data on Spain's unemployment rate. What could you do if you wanted your plot to show only dates where both Spain and Portugal have a defined unemployment rate?